### PR TITLE
Battle 2k3 2: Implement 2k3 battler placement

### PR DESCRIPTION
**Pull request by fmatthew5876**
Depends on #2217 

Fix: #2141 
Fix: #278 

Battle Conditions
----------------
- [x] Convert surround to initiative when conditions not met
- [x] Convert pincer to back when conditions not met
    - [x] Handle hidden enemies
- [x] Use correct percentage for `Troop::appear_randomly`

Placement
----------
- [x] Manual enemy placement
- [x] Enemy placement for Normal / Initiative
- [x] Enemy placement for Back
- [x] Enemy placement for surround
- [x] Enemy placement for pincer
- [x] Double check edge cases for hidden enemies in pincer / surround
- [x] Verify clamping behavior on edges of screen for enemies - enemies clamp in Y only
- [x] Verify clamping behavior on edges of screen for enemies when they transform to different enemy - nothing happens
- [x] Manual actor placement
- [x] Actor placement for Normal / Initiative
- [x] Actor placement for Back
- [x] Actor placement for surround
- [x] Actor placement for pincer
- [x] Verify clamping behavior on edges of screen for actors
- [x] Verify clamping behavior on edges of screen for actors if placed on the edge and back row pushes them out of the screen

Facing Direction
------------------
- [x] Enemy initial facing direction on battle start
- [x] Actor initial facing direction on battle start
- [x] Enemies turn to face actors once when battle begins
    - [x] Test complex manual placement
- [x] Actors turn to face enemies once when battle begins
    - [x] Test complex manual placement
- [x] Actor changes direction to attack enemy
- [x] Actor changes direction to use skill on enemy
- [x] Actor does not change direction to use skill on ally
- [x] Action does not change direction on action which targets all enemies
- [x] Enemies never change direction after battle startup

Row
----
- [x] Logic which forces back actors to the front row if all front row actors incapacitated at start of battle.
- [x] Full row implementation TDB in later PR

Things to investigate
- [x] Placement depends on battle sprite sizes. Emscripten async issues?
- [ ] Check how animation battle sprites (i.e.) "large battlers" behave

### CODE REVIEW

**Comment by fmatthew5876** [comment]
Functionality wise, I'm cutting this PR here. We now handle battler placement, facing direction, and row initialization.

The rest of row implementation will be in a follow up PR.

Once the dependencies are merged I'll rebase this again and retest.

**Comment by Ghabry** [comment]
Can you give some addresses? Want to verify this asm-wise.

**Comment by fmatthew5876** [comment]
DynRPG Info:

| Address | Class | Name |
| -- | -- | -- |
| 004977A4 | SceneBattle | Init() - parent function which does battle initialization sequence |
| 004B4910 | SceneBattle | InitActors() - initializes actor position and mirrored flag |
| 004BD1D0 | SceneBattle | InitActorFacing() - conditions for monster starting facing direction |
| 004BD1E8 | SceneBattle | ActorShouldMirror() - determine if actor should mirror |
| 004BE8E0 | SceneBattle | InitMonsters() - filters start conditions and initialize monster positions /mirror |
| 0047FD70 | SceneBattle | InitMonstorFacing() - conditions for monster starting facing direction |

| 004B4B6C | SceneBattle | ResetActorDirections() - resets the actors facing at end of battle init sequence - this is what causes them to turn |
| 004BED2C | SceneBattle | ResetMonsterDirections() - resets the monsters facing at end of battle init sequence |

**Comment by Ghabry** [comment]
When ``DrawGauge -> gauge_w`` is 0 this results in a ``zoom_x`` of infinity. Caused by previous PR:

Ubsan reports: 
```
src/transform.h:43:4: runtime error: inf is outside the range of representable values of type 'int'
    #0 0x564b0e5fe29b in Transform::Scale(double, double) src/transform.h:43:4
    #1 0x564b0e5e37e4 in Bitmap::StretchBlit(Rect const&, Bitmap const&, Rect const&, Opacity const&) src/bitmap.cpp:651:20
    #2 0x564b0e521cac in Window_Base::DrawGauge(Game_Battler const&, int, int) const src/window_base.cpp:333:12
    #3 0x564b0ed64656 in Window_BattleStatus::RefreshGauge() src/window_battlestatus.cpp:155:5
    #4 0x564b0ed61546 in Window_BattleStatus::Refresh() src/window_battlestatus.cpp:94:2
    #5 0x564b0ebd2306 in Scene_Battle_Rpg2k3::SetupSystem2Graphics() src/scene_battle_rpg2k3.cpp:167:23
    #6 0x564b0ebd1399 in Scene_Battle_Rpg2k3::OnSystem2Ready(FileRequestResult*) src/scene_battle_rpg2k3.cpp:150:2
```

**Comment by fmatthew5876** [comment]
One thing I haven't resolved is how to handle async io for web.

We have this sequence of events:
1. Create battler sprites
2. Compute battler positions based on those sprite dimensions
3. Recompute battler sprite Z values based on new battler Y values
4. Draw transition in frame containing all the sprites battlers and their Z values.
5. Perform transition in from black.

This means we have an async I/O dependency between (1) and (2). 

Also, it looks like we always have this issue with transitions. A scene start function will create the sprites (4), and then immediately TransitionIn() (4) is called in the same frame, which tries to render the transition in frame before the sprites may be loaded.

Am I overlooking something. How is it that our transition in effects on emscripten don't always look like crap?

It seems like at a minimum at least we could always wait for all graphics files when the screen is erased. A few extra frames spent spinning on a black screen won't be noticed as much as objects popping into the scene at random.

Even if we do that, I need another async hook to divide steps (1) and (2) above.

Thoughts?

**Comment by Ghabry** [comment]
At least the TransitionIn is delayed when any files are pending, e.g. when loading a new map. So most assets are already there before the transisionIn happens so they look correct. Didn't you add this "wait for In" stuff? Is is broken right now?

When the position computation depends on the dimension this needs to be delayed until the async request went through then, yes.

**Comment by fmatthew5876** [comment]
> At least the TransitionIn is delayed when any files are pending, e.g. when loading a new map. So most assets are already there before the transisionIn happens so they look correct. Didn't you add this "wait for In" stuff? Is is broken right now?

This is not the case. The `screen2` for transition in is snapped in `Transition::Init()` which occurs at the same frame as `Scene::Start()`. 

We need to wait for all graphics to finish loading, and then capture `screen2`. 

This can be seen very easily with your `EP_DEBUG_SIMULATE_ASYNC` define.

#2217 is trying to fix this.

**Comment by fmatthew5876** [comment]
I've rebased this onto #2217 and added the async code to make placement worth correctly with async I/O.

This is ready once #2217 goes in

File: src/window_base.cpp [hunk 1]
```diff
@@ -320,15 +319,18 @@ void Window_Base::DrawGauge(const Game_Battler& actor, int cx, int cy) const {
...
 	contents->Blit(cx + 0, cy, *system2, gauge_left, 255);
 	contents->Blit(cx + 16 + 25, cy, *system2, gauge_right, 255);
-
 	contents->StretchBlit(dst_rect, *system2, gauge_center, 255);
-	contents->StretchBlit(bar_rect, *system2, gauge_bar, 255);
+
+	const auto atb = actor.GetAtbGauge();
+	if (atb > 0) {
+		// Full or not full bar
+		int gauge_w = 25 * atb / actor.GetMaxAtbGauge();
```
Comment (Ghabry) [inline src/window_base.cpp:331]: This still div/0 in StretchBlit: gauge_w can become 0 here.
Then zoom_x is inf: ``src_rect.width  / dst_rect.width``

Comment (fmatthew5876) [reply src/window_base.cpp:331]: Are you sure? This logic is skipped if atb is 0 and `GetMaxAtbGauge()` is a constexpr.

Comment (Ghabry) [reply src/window_base.cpp:331]: 25 * 280 / 300000 is still 0

Comment (fmatthew5876) [reply src/window_base.cpp:331]: 🤦

**Pull request merged**